libc/string_errsig.c

```diff
@@ -0,0 +1,268 @@
+{
+/* strerror_r() not yet implemented in Phoenix-RTOS */
+#ifdef __phoenix__
+	TEST_IGNORE();
+#else
+	char buf[MAX_LEN_STRING];
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, strerror_r(0, buf, sizeof(buf)));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+#endif
+}
+
+
+TEST(string_errsingn, strerror_r_invaild)
+{
+/* strerror_r() not yet implemented in Phoenix-RTOS */
+#ifdef __phoenix__
+	TEST_IGNORE();
+#else
+	const int values[] = { INT_MIN, INT_MIN / 2, INT_MIN / 4, -1024, -256, 256, 1024, INT_MAX / 4, INT_MAX / 2, INT_MAX };
+	for (int i = 0; i < (sizeof(values) / sizeof(values[0])); i++) {
+		char buf[MAX_LEN_STRING] = { 0 };
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(EINVAL, strerror_r(values[i], buf, sizeof(buf)));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+	}
+#endif
+}
+
+
+TEST(string_errsingn, strerror_r_overflow)
+{
+/* strerror_r() not yet implemented in Phoenix-RTOS */
+#ifdef __phoenix__
+	TEST_IGNORE();
+#else
+	char buf[1];
+	char buf2[MAX_LEN_STRING];
+	for (int i = 0; i < error_codes_len; i++) {
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(ERANGE, strerror_r(error_codes[i], buf, sizeof(buf)));
+		TEST_ASSERT_EQUAL_INT(ERANGE, strerror_r(error_codes[i], buf2, 2));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+	}
+#endif
+}
+
+
+TEST(string_errsingn, strsignal_basic)
+{
+	for (int i = 0; i < signal_codes_len; i++) {
+		errno = 0;
+		TEST_ASSERT_NOT_NULL(strsignal(signal_codes[i]));
+
+		if (i != 0) {
+			TEST_ASSERT_NOT_NULL(strsignal(signal_codes[i - 1]));
+			TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(strsignal(signal_codes[i - 1]), strsignal(signal_codes[i])));
+		}
+
+		TEST_ASSERT_EQUAL_INT(0, errno);
+	}
+}
+
+
+TEST(string_errsingn, strsignal_invaild)
+{
+	const int values[] = { INT_MIN, INT_MIN / 2, INT_MIN / 4, -1024, -256, -128, -64, -1, 0, 128, 256, 1024, INT_MAX / 4, INT_MAX / 2, INT_MAX };
+	for (int i = 0; i < (sizeof(values) / sizeof(values[0])); i++) {
+		errno = 0;
+		TEST_ASSERT_NOT_NULL(strsignal(signal_codes[i]));
+		TEST_ASSERT_NOT_NULL(strstr(strsignal(values[i]), "Unknown signal"));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+	}
+}
+
+
+TEST(string_errsingn, strsignal_real_time)
+{
+/* Disabled because of #687 issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/687 */
+#ifdef __phoenix__
+	TEST_IGNORE_MESSAGE("#687 issue");
+#else
+	TEST_ASSERT_EQUAL_STRING("Real-time signal 0", strsignal(SIGRTMIN));
+	TEST_ASSERT_EQUAL_STRING("Real-time signal 1", strsignal(SIGRTMIN + 1));
+	TEST_ASSERT_EQUAL_STRING("Real-time signal 16", strsignal(SIGRTMIN + RTSIG_MAX / 2));
+	TEST_ASSERT_EQUAL_STRING("Unknown signal 66", strsignal(SIGRTMIN + RTSIG_MAX));
+
+	TEST_ASSERT_EQUAL_STRING("Real-time signal 30", strsignal(SIGRTMAX));
+	TEST_ASSERT_EQUAL_STRING("Real-time signal 29", strsignal(SIGRTMAX - 1));
+	TEST_ASSERT_EQUAL_STRING("Real-time signal 14", strsignal(SIGRTMAX - RTSIG_MAX / 2));
+	TEST_ASSERT_EQUAL_STRING("Unknown signal 32", strsignal(SIGRTMAX - RTSIG_MAX));
+#endif
+}
+
+
+TEST(string_errsingn, strsignal_psignal)
```

Comment:
I think we can split psignal tests to a few cases, like for the functions before (not null checks and so on) and then add a case where we compare outputs of strerror and psignal. Now it would be hard to investigate what was the reason of fail, the wrong output/error of psignal or the fact that outputs for strsignal and psignal are different. This way we will assure that there are no basic errors with those functions and then we will comapre their messages in more detailed test case.